blivet/osinstall.py

```diff
@@ -1340,10 +1359,93 @@ def _update_custom_storage_ksdata(self):
         self.fsset = FSSet(self.devicetree)
 
     def reset(self, cleanup_only=False):
+        """ Reset storage configuration to reflect actual system state.
+
+            This will cancel any queued actions and rescan from scratch but not
+            clobber user-obtained information like passphrases, iscsi config, &c
+
+            :keyword cleanup_only: prepare the tree only to deactivate devices
+            :type cleanup_only: bool
+
+            See :meth:`devicetree.Devicetree.populate` for more information
+            about the cleanup_only keyword argument.
+        """
+        log.info("resetting Blivet (version %s) instance %s", __version__, self)
+        if flags.installer_mode:
+            # save passphrases for luks devices so we don't have to reprompt
+            self.encryption_passphrase = None
+            for device in self.devices:
+                if device.format.type == "luks" and device.format.exists:
+                    self.save_passphrase(device)
+
+        if self.ksdata:
+            self.config.update(self.ksdata)
+
+        if flags.installer_mode and not flags.image_install:
+            iscsi.startup()
+            fcoe.startup()
+            zfcp.startup()
+
         super().reset(cleanup_only=cleanup_only)
 
         self.fsset = FSSet(self.devicetree)
 
+        # protected device handling
+        self.protected_dev_names = []
+        self._resolve_protected_device_specs()
+        self._find_live_backing_device()
+        for devname in self.protected_dev_names:
+            dev = self.devicetree.get_device_by_name(devname)
+            self._mark_protected_device(dev)
+
+        self.roots = []
+        if flags.installer_mode:
+            self.roots = find_existing_installations(self.devicetree)
+            self.dump_state("initial")
+
+        if not flags.installer_mode:
+            self.devicetree.handle_nodev_filesystems()
```

Comment:
At least the protected device handling can't go back to Blivet.reset, since self.protected_dev_specs is part of StorageDiscoveryConfig, which doesn't get instantiated in Blivet anymore. The other stuff (self.roots = [] onward) I can move, though.